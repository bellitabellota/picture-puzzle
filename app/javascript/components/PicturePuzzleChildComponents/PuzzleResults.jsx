import React, {useEffect, useState} from "react";
import { useParams, Link } from "react-router-dom";

function PuzzleResults() {
  const params = useParams();
  const [puzzleResults, setPuzzleResults] = useState([]);
  const [puzzleTitle, setPuzzleTitle] = useState(null);

  useEffect(() => {
    const url =`/api/v1/picture_puzzles/${params.id}/results`;
    fetch(url)
    .then((response) => {
      if(!response.ok) {
        throw new Error("Network response not ok")
      }
      return response.json()
    }).then((data) => {console.log(data.puzzleTitle)

      setPuzzleTitle(data.puzzleTitle);
      setPuzzleResults(data.results);
    }).catch((error) => {
      console.log(error);
    })
  }, [])

  const resultElements = puzzleResults.map((result, index) => {
    const minutes = Math.floor(result.seconds_to_completion/ 60);
    const remainingSeconds = result.seconds_to_completion % 60;

    return(
      <div key={result.id}>
        <p className="rank">{index + 1}</p>
        <p className="name">{result.name}</p>
        <p className="duration">{`${minutes} m ${remainingSeconds} s `}</p>

      </div>
    )
  })

  return (
    <main className="main-picture-puzzle-results">

    <Link to="/" >&lt; Back to Home</Link>
    <h1>{puzzleTitle} - Results</h1>

    <div className="results-container">
      <div className="label-container">
        <p className="rank-label">Rank</p>
        <p className="name-label">Name</p>
        <p className="duration-label">Duration</p>
      </div>
      { resultElements }
    </div>
  </main>
  )
}

export default PuzzleResults;