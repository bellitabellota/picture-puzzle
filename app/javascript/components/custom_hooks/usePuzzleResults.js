import {useEffect, useState} from "react";

const usePuzzleResults = (paramsId) => {
  const [puzzleResults, setPuzzleResults] = useState([]);
  const [puzzleTitle, setPuzzleTitle] = useState(null);

  useEffect(() => {
    const url =`/api/v1/picture_puzzles/${paramsId}/results`;
    fetch(url)
    .then((response) => {
      if(!response.ok) {
        throw new Error("Network response not ok")
      }
      return response.json()
    }).then((data) => {
      setPuzzleTitle(data.puzzleTitle);
      setPuzzleResults(data.results);
    }).catch((error) => {
      console.log(error);
    })
  }, [])

  return {puzzleResults, puzzleTitle}

}

export default usePuzzleResults;