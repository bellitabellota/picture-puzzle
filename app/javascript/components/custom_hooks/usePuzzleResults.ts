import {useEffect, useState} from "react";

const usePuzzleResults = (paramsId: string) => {
  const [puzzleResults, setPuzzleResults] = useState([]);
  const [puzzleTitle, setPuzzleTitle] = useState(null);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    const url =`/api/v1/picture_puzzles/${paramsId}/results`;
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error(`HTTP Error ${response.status}: ${response.statusText}`)
      }
      return response.json()
    }).then((data) => {
      setPuzzleTitle(data.puzzleTitle);
      setPuzzleResults(data.results);
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    }).finally(() => {
      if (!signal.aborted) {
        setIsLoading(false);
      }
    })

    return () => controller.abort(); // the cleanup function was added because the effect runs when the component mounts. So in case the component unmount before the request completes, the request would have been still active but the component would not be there anymore to handle the response.
  }, [])

  return {puzzleResults, puzzleTitle, error, isLoading}

}

export default usePuzzleResults;