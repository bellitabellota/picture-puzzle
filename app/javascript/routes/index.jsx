import PicturePuzzles from "../components/PicturePuzzles";
import PicturePuzzle from "../components/PicturePuzzle"
import React from "react";

const routes = [{
  path: "/",
  element: <PicturePuzzles />
}, {
  path: "/:id",
  element: <PicturePuzzle />
}]

export default routes;