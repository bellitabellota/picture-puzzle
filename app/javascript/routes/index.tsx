import PicturePuzzles from "../components/PicturePuzzles";
import PicturePuzzle from "../components/PicturePuzzle"
import PuzzleResults from "../components/PuzzleResults";
import React from "react";

const routes = [{
  path: "/",
  element: <PicturePuzzles />
}, {
  path: "/:id",
  element: <PicturePuzzle />
}, {
  path: "/:id/results",
  element: <PuzzleResults />
}]

export default routes;