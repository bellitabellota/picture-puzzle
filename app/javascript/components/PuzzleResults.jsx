import React, {useEffect, useState} from "react";
import { useParams, Link } from "react-router-dom";
import usePuzzleResults from "./custom_hooks/usePuzzleResults";

function PuzzleResults() {
  const params = useParams();
  const {puzzleResults, puzzleTitle} = usePuzzleResults(params.id);

  const resultElements = puzzleResults.map((result, index) => {
    const minutes = Math.floor(result.seconds_to_completion/ 60);
    const remainingSeconds = result.seconds_to_completion % 60;

    return(
      <div key={result.id}>
        <p className="rank">{index + 1}</p>
        <p className="name">{result.name}</p>
        <p className="duration">{`${minutes} m ${remainingSeconds} s `}</p>
      </div>
    )
  })

  return (
    <main className="main-picture-puzzle-results">

    <Link to="/" >&lt; Back to Home</Link>
    <h1>{puzzleTitle} - Results</h1>

    <div className="results-container">
      <div className="label-container">
        <p className="rank-label">Rank</p>
        <p className="name-label">Name</p>
        <p className="duration-label">Duration</p>
      </div>
      { resultElements }
    </div>
  </main>
  )
}

export default PuzzleResults;