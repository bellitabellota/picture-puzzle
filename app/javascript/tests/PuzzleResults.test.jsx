import React from "react";
import {render, screen } from "@testing-library/react";
import PuzzleResults from "../components/PuzzleResults";
import usePuzzleResults from "../components/custom_hooks/usePuzzleResults";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import routes from "../routes/index"


vi.mock("../components/custom_hooks/usePuzzleResults", () => ({
  default: vi.fn(),
}))

describe("PuzzleResults", () => {
  it("renders 'Puzzle Results are loading ...' if isLoading is true", () => {
    usePuzzleResults.mockReturnValue({puzzleResults: [], puzzleTitle: null,  error: null, isLoading: true})
    render(<PuzzleResults />)

    expect(screen.getByText("Puzzle Results are loading ...")).toBeInTheDocument();
  })

  it("renders error message if sePuzzleResults returned an Error", () => {
    usePuzzleResults.mockReturnValue({puzzleResults: [], puzzleTitle: null, error: new Error("server error"), isLoading: false})

    render(<PuzzleResults />)
    expect(screen.getByText("server error")).toBeInTheDocument();
  })

})