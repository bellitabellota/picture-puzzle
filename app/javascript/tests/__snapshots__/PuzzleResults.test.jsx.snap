// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PuzzleResults > renders puzzleResults and puzzleTitle if usePuzzleResults returned puzzleResults 1`] = `
<div>
  <main
    class="main-picture-puzzle-results"
  >
    <a
      data-discover="true"
      href="/"
    >
      &lt; Back to Home
    </a>
    <h1>
      Test Puzzle 123
       - Results
    </h1>
    <div
      class="results-container"
    >
      <div
        class="label-container"
      >
        <p
          class="rank-label"
        >
          Rank
        </p>
        <p
          class="name-label"
        >
          Name
        </p>
        <p
          class="duration-label"
        >
          Duration
        </p>
      </div>
      <div>
        <p
          class="rank"
        >
          1
        </p>
        <p
          class="name"
        >
          Player 1
        </p>
        <p
          class="duration"
        >
          0 m 28 s 
        </p>
      </div>
      <div>
        <p
          class="rank"
        >
          2
        </p>
        <p
          class="name"
        >
          Player 2
        </p>
        <p
          class="duration"
        >
          0 m 35 s 
        </p>
      </div>
      <div>
        <p
          class="rank"
        >
          3
        </p>
        <p
          class="name"
        >
          Player 3
        </p>
        <p
          class="duration"
        >
          1 m 8 s 
        </p>
      </div>
    </div>
  </main>
</div>
`;
